File: 2020/thanksgiving-contest/08 Grocery Shopping on a Budget/grocery.cpp
// My solution

#include <bits/stdc++.h>

using namespace std;

int main()
{
    int n, m;
    cin >> n >> m;
    int travel[n];
    for(int i = 0; i<n; i++) {
        cin >> travel[i];
    }
    int grocery[n][m];
    for(int i = 0; i<n; i++) {
        for(int j = 0; j<m; j++) {
            cin >> grocery[i][j];
        }
    }
    long long cost, ans=-1;
    int items[m]; //min cost for each item
    for(int i = 1; i < (1<<n); i++) { //generate all subsets of stores
        cost=0;
        for(int j = 0; j < m; j++) {
            items[j]=-1;
        }
        for(int j = 0; j < n; j++) {
            if(i&(1<<j)) { //travel to store j?
                cost+=travel[j];
                for(int k = 0; k < m; k++) {
					//update min cost
                    if(items[k]==-1) {
                        items[k]=grocery[j][k];
                    } else {
                        items[k]=min(items[k], grocery[j][k]);
                    }
                }
            }
        }
        for(int j = 0; j<m; j++) {
            cost+=items[j];
        }
        if(ans==-1) {
            ans=cost;
        } else {
            ans=min(ans, cost);
        }
    }
    cout << ans << endl;
}

File: 2020/thanksgiving-contest/08 Grocery Shopping on a Budget/grocery.py
# Submitted by Nathan Dai
# The C++ solution may be more understandable, even if you don't know C++

from itertools import combinations, chain

class Store:
    def __init__(self, gas: int, items: list):
        self.gas = gas
        self.items = items
    
    def __repr__(self) -> str:
        return str(self.gas) + ' : ' + str(self.items)

n, m = map(int, input().split())
gas = list(map(int, input().split()))
stores = list(zip(gas, [list(map(int, input().split())) for _ in range(n)]))
stores = [Store(i[0], i[1]) for i in stores]

cheapest = float('inf')
for combination in chain(*[list(combinations(stores, i)) for i in range(1, n + 1)]):
    price = sum([store.gas for store in combination])
    price += sum([min(grocery) for grocery in zip(*[j.items for j in combination])])

    if (price < cheapest): 
        cheapest = price

print(cheapest)
